webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/DeploymentService.scala

```diff
@@ -335,113 +335,132 @@ trait PromiseGenerationService {
-      _                     =  PolicyLogger.debug(s"Node contexts built in ${timeNodeContexts} ms, start to build new node configurations.")
-
-      buildConfigTime       =  System.currentTimeMillis
-      /// here, we still have directive by directive info
-      configsAndErrors      <- buildNodeConfigurations(activeNodeIds, ruleVals, nodeContexts, allNodeModes, scriptEngineEnabled, globalPolicyMode, parallelism, jsTimeout, generationContinueOnError) ?~! "Cannot build target configuration node"
-      /// only keep successfull node config. We will keep the failed one to fail the whole process in the end if needed
-      nodeConfigs           =  configsAndErrors.ok.map(c => (c.nodeInfo.id, c)).toMap
-      timeBuildConfig       =  (System.currentTimeMillis - buildConfigTime)
-      _                     =  PolicyLogger.debug(s"Node's target configuration built in ${timeBuildConfig} ms, start to update rule values.")
-
-      updatedNodeConfigIds  =  getNodesConfigVersion(nodeConfigs, nodeConfigCaches, generationTime)
-      updatedNodeConfigs    =  nodeConfigs.filterKeys(id => updatedNodeConfigIds.keySet.contains(id))
-
-      reportTime            =  System.currentTimeMillis
-      expectedReports       =  computeExpectedReports(ruleVals, updatedNodeConfigs.values.toSeq, updatedNodeConfigIds, generationTime, allNodeModes)
-      timeSetExpectedReport =  (System.currentTimeMillis - reportTime)
-      _                     =  PolicyLogger.debug(s"Reports updated in ${timeSetExpectedReport} ms")
-
-      ///// so now we have everything for each updated nodes, we can start writing node policies and then expected reports
-
-      // WHY DO WE NEED TO FORGET OTHER NODES CACHE INFO HERE ?
-      _                     <- forgetOtherNodeConfigurationState(nodeConfigs.keySet) ?~! "Cannot clean the configuration cache"
-
-      writeTime             =  System.currentTimeMillis
-      writtenNodeConfigs    <- writeNodeConfigurations(rootNodeId, updatedNodeConfigIds, nodeConfigs, allLicenses, globalPolicyMode, generationTime, parallelism) ?~!"Cannot write nodes configuration"
-      timeWriteNodeConfig   =  (System.currentTimeMillis - writeTime)
-      _                     =  PolicyLogger.debug(s"Node configuration written in ${timeWriteNodeConfig} ms, start to update expected reports.")
+      (updatedNodesId, updatedNodeInfo, expectedReports, allErrors, errorNodes, timeFetchAll, timeHistorize, timeRuleVal, timeBuildConfig, timeWriteNodeConfig, timeSetExpectedReport) <- for {
+        (updatedNodeConfigIds, allNodeConfigurations, allNodeConfigsInfos, updatedNodesId, updatedNodeInfo, allLicenses, globalPolicyMode, allNodeModes, allErrors, errorNodes, timeFetchAll, timeHistorize, timeRuleVal, timeBuildConfig) <- for {
+          (activeNodeIds, ruleVals, nodeContexts, allNodeModes, scriptEngineEnabled, globalPolicyMode, nodeConfigCaches, allLicenses, timeFetchAll, timeHistorize, timeRuleVal) <- for {
```

Comment:
The scoping is quickly becoming hard to follow I believe. Wouldn't it be clearer to factor out the inner-most `for` in dedicated support function ? (I'm not sure, because we already have tons of them, and we will likely loose the general understanding of the policy generation workflow). What's your opinion on that?

Reply:
Or perhaps build dedicated `case` objects that only handle the relevant part and are passed around ? 
(well, I fear that it will make refactoring harder, since we will have a hard time to know when we can remove an object from such a structure, so perhaps a bad idea actually)

Reply:
we already have a lot of back and forth with def, making it hard to follow. Plus, there are timestamp for loging within this, so adding them in a def is nightmarish

Reply:
I wish there were a way to make it better (an annotation that the compiler could use to know to release quickly some variable)